packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx

```diff
@@ -199,13 +203,148 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
         });
     }
 
+    /**
+     * @param filePath Path of the file.
+     * @param rootUris URIs of the root folders in the current workspace.
+     * @returns URI of the root folder that a file is belong to.
+     * If a file belongs to more than one root folder, return the root folder closest to the file.
+     * If the file is not from the current workspace, return an empty URI.
+     */
+    getRoot(filePath: string, rootUris: string[]): URI {
+        const roots = rootUris.filter(root => new URI(root).withScheme('file').isEqualOrParent(new URI(filePath).withScheme('file')));
+        if (roots.length > 0) {
+            return new URI(roots.sort((r1, r2) => r2.length - r1.length)[0]);
+        }
+        return new URI();
+    }
+
+    /**
+     * Find matches in a single widget.
+     * @param searchTerm  The search keyword.
+     * @param widget The widget that is being searched.
+     * @param fileContent The content of the widget.
+     * @param searchOptions The options for search operation.
+     */
+    findMatches(searchTerm: string, widget: EditorWidget, fileContent: string, searchOptions: SearchInWorkspaceOptions): SearchMatch[] {
+        let matches: SearchMatch[] = [];
+        const allLines: string[] = fileContent.split(/[\n\u0085\u2028\u2029]|\r\n?/);
+        const findLineText = function (lineNum: number): string {
+            return allLines[--lineNum];
+        };
+        if (widget.editor instanceof MonacoEditor) {
+            const model = widget.editor.getControl().getModel();
+            const results: monaco.editor.FindMatch[] | undefined = model?.findMatches(
+                searchTerm,
+                true,
+                !!searchOptions.useRegExp,
+                !!searchOptions.matchCase,
+                // eslint-disable-next-line no-null/no-null
+                searchOptions.matchWholeWord ? searchTerm : null,
+                true,
+                searchOptions.maxResults);
```

Comment:
I'm not sure about limiting with `maxResults`, this number is used to get limit the search results obtained by `ripgrep`.
If dirty editors consume results, we should make sure to sum their result count and limit the overall results after.

The other problem is that `ripgrep` will produce duplicate results (for the same dirty editor URI), and the `maxResults` will also be affected.

Reply:
For this issue, currently I'm having a workaround by substracting `maxResults` directly in `monaco find matches` (results from monaco should be prioritized). In addition, I added a check for duplicate results from `ripgrep`, and increment `maxResults` by 1 to reimburse the unwanted loss of result counter.